Beiwe/Recline/Recline.swift

```diff
@@ -31,175 +29,158 @@ class Recline {
             return _save(obj)
         }
     }
-    
+
     func _save<T: ReclineObject>(_ obj: T) {
-        guard let db = db else {
-            fatalError("again the database is not instantiated - what do you think you are doing?")
+        // give up early if db is not instantiated
+        guard let db = self.db else {
+            fatalError(
+                "again the database is not instantiated - what do you think you are doing?"
+            )
+        }
+        // get or create the document.  I can't tell if this could be made non-optional,
+        var doc: MutableDocument?
+        if let _id = obj._id {
+            do {
+                doc = try db.defaultCollection().document(id: _id)?.toMutable()
+            } catch {
+                fatalError("no doc found")
+            }
+        } else {
+            doc = MutableDocument()
         }
-        // get or create the document.
-        let doc: CBLDocument = if let _id = obj._id { db.document(withID: _id)! } else { db.createDocument() }
-        
         // take the object, make a json version of it, and .... then we do something involving ReclineMetadata?
         var newProps: [String: Any] = Mapper<T>().toJSON(obj)
-        let reclineMeta = ReclineMetadata(type: String(describing: type(of: obj)))
-        newProps[kReclineMetadataKey] = Mapper<ReclineMetadata>().toJSON(reclineMeta)
-        
-        // couchbase uses a revisioning system for concurrency(?), so this is probably assigning an id and a revision
-        // these are nil on registration (e.g. when the database is empty), need the ?
-        newProps["_id"] = doc.properties?["_id"]
-        newProps["_rev"] = doc.properties?["_rev"]
-        
-        // I think we are swallowing any errors
+        let reclineMeta = ReclineMetadata(
+            type: String(describing: type(of: obj)))
+        newProps[kReclineMetadataKey] = Mapper<ReclineMetadata>().toJSON(
+            reclineMeta)
+
+        for (key, value) in newProps {
+            doc?.setValue(value, forKey: key)
+        }
+
         do {
-            try doc.putProperties(newProps) // commit to database
+            if let doc = doc {
+                try db.defaultCollection().save(document: doc)
+                print(
+                    "Created document id type \(doc.id)? with patientId = \(doc.string(forKey: "patientId")!)"
+                )
+            } else {
+                fatalError("Failed to create or fetch the document for saving")
+            }
         } catch {
-            fatalError("error saving to couchbase? \(error)")
+            fatalError("cannot save document")
         }
     }
-    
+
     /// gets all rows in the typesView database view?
     // runs the query across everything and returns the contents.
     func queryAll<T: ReclineObject>() -> [T] {
         OUTER_RECLINE_QUEUE.sync {
             return self._queryAll()
         }
     }
-    
+
     func _queryAll<T: ReclineObject>() -> [T] {
```

Comment:
I need this queryAll function explained - I understand virtually nothing about this file, it was crap before I started trying to clean stuff up and fixing the crashes.

Reply:
Well, basically you start with creating a query object which will return results with the document ID and all the keys/columns. 

Then, the result of the executed query object is `guarded` to make sure that we have `docId` and the document data itself (a `DictionaryObject`).

The rest of the code is basically mapping the `DictionaryObject` into `ReclineObject` type. I believe the mapping complexity here was not necessary. But, I did not want to spend more time simplifying it. I was simply reimplementing what was already there.